core/src/main/java/org/elasticsearch/search/fetch/subphase/CollapseFetchSubPhase.java

```diff
@@ -0,0 +1,187 @@
+import java.util.HashMap;
+
+/**
+ * A fetch sub phase that expands the top documents if they were collapsed by
+ * {@link org.apache.lucene.search.grouping.CollapsingTopDocsCollector}.
+ * For each group the top N documents are retrieved using the internal {@link Client} to perform a distributed query
+ * in the cluster.
+ */
+public class CollapseFetchSubPhase implements FetchSubPhase {
+    @Override
+    public void hitsExecute(SearchContext context, InternalSearchHit[] hits) {
+        if (context.collapse() == null) {
+            return ;
+        }
+        hits = hits.clone(); // don't modify the incoming hits
+        Arrays.sort(hits, (a, b) -> Integer.compare(a.docId(), b.docId()));
+
+        SortField collapseSortField = context.collapse().getSortField();
+        boolean isRelevanceOnly = false;
+        if (context.sort() == null || context.sort().sort == Sort.RELEVANCE) {
+            isRelevanceOnly = true;
+        }
+
+        try {
+            final CollapsingFieldSource source = new CollapsingFieldSource(collapseSortField);
+            final IndexReader indexReader = context.searcher().getIndexReader();
+            int readerIndex = -1;
+            int docBase = 0;
+            for (InternalSearchHit hit : hits) {
+                // Retrieve the collapse value
+                int hitReaderIndex = ReaderUtil.subIndex(hit.docId(), indexReader.leaves());
+                if (readerIndex != hitReaderIndex) {
+                    readerIndex = hitReaderIndex;
+                    LeafReaderContext ctx = indexReader.leaves().get(readerIndex);
+                    docBase = ctx.docBase;
+                    source.setNextReader(ctx);
+                }
+                int docID = hit.docId() - docBase;
+                Object collapseValue = source.get(docID);
+                // Set the metadata field _collapse
+                if (hit.fieldsOrNull() == null) {
+                    hit.fields(new HashMap<>(2));
+                }
+                SearchHitField hitField = new InternalSearchHitField("_collapse", new ArrayList<>(1));
+                if (collapseValue != null) {
+                    String collapseTerm = context.collapse().getFieldType().valueForDisplay(collapseValue).toString();
+                    hitField.values().add(collapseTerm);
+                } else {
+                    hitField.values().add(null);
+                }
+                hit.fields().put("_collapse", hitField);
+
+                if (isRelevanceOnly) {
+                    // Reset the sort values, only _score is needed
+                    hit.sortValues(new Object[0], new DocValueFormat[0]);
+                }
+
+                // Retrieve the inner_hits for this group
+                InnerHitBuilder options = context.collapse().getInnerHit();
+                if (options == null) {
+                    continue;
+                }
+                Client client = context.getQueryShardContext().getClientNoCache();
```

Comment:
I don't think we should make external calls here. This is the wrong place to do this.

We should do this at a different level. I think the best place is a hook point that @polyfractal worked in #19587. On the transport layer it is okay to make additional external calls. Unfortunately Zach's PR never got merged. 

What we can do is move this code that executes a search and enriches the search response to the same place Zach tried to add this extension point.

I think we should add something like this to `SearchTransportAction.java`:

At line 164

```diff
Index: core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java	(date 1483612278000)
+++ core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java	(revision )
@@ -33,15 +33,26 @@
 import org.elasticsearch.common.settings.Setting.Property;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.Index;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.InnerHitBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.query.TermQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
 import org.elasticsearch.search.SearchService;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.collapse.CollapseBuilder;
 import org.elasticsearch.search.internal.AliasFilter;
+import org.elasticsearch.search.internal.InternalSearchHit;
+import org.elasticsearch.search.sort.SortBuilder;
 import org.elasticsearch.tasks.Task;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.TransportService;
 
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Executor;
@@ -150,8 +161,23 @@
             }
         }
 
+        // Only enrich the search response iff collapsing has been specified:
+        final ActionListener<SearchResponse> wrapper;
+        if (searchRequest.source().collapse() != null && searchRequest.source().collapse().getInnerHit() != null) {
+            wrapper = ActionListener.wrap(searchResponse -> {
+                if (searchResponse.getHits().getHits().length == 0) {
+                    listener.onResponse(searchResponse);
+                } else {
+                    Iterator<SearchHit> hitsIterator = searchResponse.getHits().iterator();
+                    expandCollapsedHits(searchRequest, hitsIterator, searchResponse, listener);
+                }
+            }, listener::onFailure);
+        } else {
+            wrapper = listener;
+        }
+
         searchAsyncAction((SearchTask)task, searchRequest, shardIterators, startTimeInMillis, clusterState,
-            Collections.unmodifiableMap(aliasFilter), concreteIndexBoosts, listener).start();
+            Collections.unmodifiableMap(aliasFilter), concreteIndexBoosts, wrapper).start();
     }
 
     @Override
@@ -204,4 +230,82 @@
                 + "] to a greater value if you really want to query that many shards at the same time.");
         }
     }
+
+    void expandCollapsedHits(SearchRequest originalSearchRequest, Iterator<SearchHit> hits, SearchResponse searchResponse, ActionListener<SearchResponse> finalListener) {
+        if (hits.hasNext() == false) {
+            finalListener.onResponse(searchResponse);
+            return;
+        }
+
+        CollapseBuilder collapseBuilder = originalSearchRequest.source().collapse();
+        InternalSearchHit searchHit = (InternalSearchHit) hits.next();
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder();
+        Object collapseValue = searchHit.field("_collapse").getValue();
+        if (collapseValue != null) {
+            boolQuery.filter(new TermQueryBuilder(collapseBuilder.getField(), collapseValue));
+        } else {
+            boolQuery.mustNot(QueryBuilders.existsQuery(collapseBuilder.getField()));
+        }
+        QueryBuilder query = originalSearchRequest.source().query();
+        if (query != null) {
+            boolQuery.must(query);
+        }
+
+        SearchRequest searchRequest = new SearchRequest(originalSearchRequest.indices());
+        searchRequest.types(originalSearchRequest.types());
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.query(boolQuery);
+        addInnerHitOptions(sourceBuilder, collapseBuilder.getInnerHit());
+        searchRequest.source(sourceBuilder);
+        execute(searchRequest, ActionListener.wrap(innerHitsSearchResponse -> {
+            SearchHits innerHits = innerHitsSearchResponse.getHits();
+            if (searchHit.getInnerHits() == null) {
+                searchHit.setInnerHits(new HashMap<>(1));
+            }
+            searchHit.getInnerHits().put(collapseBuilder.getInnerHit().getName(), innerHits);
+            expandCollapsedHits(originalSearchRequest, hits, searchResponse, finalListener);
+        }, finalListener::onFailure));
+    }
+
+    private void addInnerHitOptions(SearchSourceBuilder source, InnerHitBuilder options) {
+        source.from(options.getFrom());
+        source.size(options.getSize());
+        if (options.getSorts() != null) {
+            for (SortBuilder<?> sort : options.getSorts()) {
+                source.sort(sort);
+            }
+        } else if (source.sorts() != null) {
+            for (SortBuilder<?> sort : source.sorts()){
+                source.sort(sort);
+            }
+        }
+        if (options.getFetchSourceContext() != null) {
+            if (options.getFetchSourceContext().includes() == null && options.getFetchSourceContext().excludes() == null) {
+                source.fetchSource(options.getFetchSourceContext().fetchSource());
+            } else {
+                source.fetchSource(options.getFetchSourceContext().includes(),
+                    options.getFetchSourceContext().excludes());
+            }
+        }
+        if (options.getDocValueFields() != null) {
+            for (String field : options.getDocValueFields()) {
+                source.docValueField(field);
+            }
+        }
+        if (options.getStoredFieldsContext() != null && options.getStoredFieldsContext().fieldNames() != null) {
+            for (String field : options.getStoredFieldsContext().fieldNames()) {
+                source.storedField(field);
+            }
+        }
+        if (options.getScriptFields() != null) {
+            for (SearchSourceBuilder.ScriptField field : options.getScriptFields()) {
+                source.scriptField(field.fieldName(), field.script());
+            }
+        }
+        if (options.getHighlightBuilder() != null) {
+            source.highlighter(options.getHighlightBuilder());
+        }
+        source.explain(options.isExplain());
+        source.trackScores(options.isTrackScores());
+    }
 }

```

Then when the work on #19587 continues, this expand collapsed hit logic should be migrated too.